### PR TITLE
feat: basic permission control

### PR DESCRIPTION
**Pull request by exuanbo**

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -4,7 +4,10 @@ import type { Awareness } from 'y-protocols/awareness'
 export type RoomName = Room
 export type ClientId = Awareness['clientID']
 
+export type DefaultClientData = {}
```
Comment (Himself65) [inline src/types.ts:7]: make this into a `interface` and downstream could re-declare and extends it

**Comment by exuanbo** [comment]
What do you think about this API design?

/cc @Himself65

File: src/server/socket/index.ts [hunk 2]
```diff
@@ -35,94 +55,155 @@ declare module 'socket.io' {
...
   })
 
+  if (autoDeleteRoom) {
+    adapter.on('delete-room', (roomName: RoomName) => {
+      if (isDefaultRoom(roomName) || !roomMap.has(roomName)) {
+        return
+      }
+      const room = roomMap.get(roomName)!
+      roomMap.delete(roomName)
+      void room.destroy()
```
Comment (Himself65) [inline src/server/socket/index.ts:142]: no void

**Review by Himself65 — CHANGES_REQUESTED** [review]

**Review by Himself65 — CHANGES_REQUESTED** [review]

File: .eslintrc.js [hunk 3]
```diff
@@ -12,8 +12,5 @@ module.exports = {
     next: {
       rootDir: 'app'
     }
-  },
-  rules: {
-    'no-void': ['error', { allowAsStatement: true }]
```
Comment (Himself65) [inline .eslintrc.js:17]: I will move this to dev-kit later

File: src/events.ts [hunk 4]
```diff
@@ -24,18 +27,25 @@ type ValidateEvents<
...
   ['doc:update']: (updateV2: ArrayBuffer) => void
   ['awareness:update']: (update: ArrayBuffer) => void
 }>
 
-export type ClientToServerEvents = ValidateEvents<{
+export interface ServerToClientEvents<ClientData extends DefaultClientData = DefaultClientData>
+  extends DefaultServerToClientEvents<ClientData> {}
+
+export type DefaultClientToServerEvents = ValidateEvents<{
+  ['room:close']: () => void
```
[the remark this answers is not in the document]
Comment (exuanbo) [reply src/events.ts:41]: This event is emitted in function `provider.closeRoom()` and it's not part of the sync protocol

**Review by Himself65 — APPROVED** [review]
There's a logic issue. 

If an owner closes a room, others could open it again with the same id and data. This caused the original owner to become a user. But we could fix it in the future.

**Comment by exuanbo** [comment]
> There's a logic issue.
> 
> If an owner closes a room, others could open it again with the same id and data. This caused the original owner to become a user. But we could fix it in the future.

I've considered this issue. It will only happen when the server is created with option `autoDeleteRoom` (default to `false`).

I enabled this option in the example because it's easier to test.

https://github.com/TexteaInc/y-socket.io/blob/985dcd0909448b9ff45889f55c472eceb76d5613/example/scripts/server.ts#L15-L17

And currently the room data (`owner` etc.) is stored in memory by `roomMap`, in the future it will be better to provide a way to store the data in external database or something.

**Pull request merged**